store.go

```diff
@@ -0,0 +1,193 @@
+// TODO document this, in particular provide some details around:
+// - the purpose and use of "buckets"
+type expirableStore struct {
+	maxSize int
+
+	items map[string]*entry
+
+	buckets            []bucket
+	bucketTTL          time.Duration
+	numberBuckets      int
+	nextBucketToExpire int
+
+	mu sync.Mutex
+
+	pool sync.Pool
+
+	cancelFunc context.CancelFunc
+	ctx        context.Context
+}
+
+func newExpireableStore(maxSize int, maxEntryTTL time.Duration, o ...Option) (*expirableStore, error) {
+	const op = "rate.newExpireableStore"
+
+	opts := getOpts(o...)
+
+	switch {
+	case maxSize <= 0:
+		return nil, fmt.Errorf("%s: max size must be greater than zero: %w", op, ErrInvalidMaxSize)
+	case maxEntryTTL <= 0:
+		return nil, fmt.Errorf("%s: max entry ttl must be greater than zero: %w", op, ErrInvalidParameter)
+	case opts.withNumberBuckets <= 0:
+		return nil, fmt.Errorf("%s: number of buckets must be greater than zero: %w", op, ErrInvalidNumberBuckets)
+	}
+
+	bucketTTL := (maxEntryTTL) / time.Duration(opts.withNumberBuckets-1)
+
+	buckets := make([]bucket, opts.withNumberBuckets)
+	for i := 0; i < opts.withNumberBuckets; i++ {
+		buckets[i] = bucket{
+			entries: make(map[string]*entry, maxSize),
+		}
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	s := &expirableStore{
+		maxSize:       maxSize,
+		items:         make(map[string]*entry, maxSize),
+		buckets:       buckets,
+		bucketTTL:     bucketTTL,
+		numberBuckets: opts.withNumberBuckets,
+		pool: sync.Pool{
+			New: func() any {
+				return &entry{
+					value: &Quota{},
+				}
+			},
+		},
+		cancelFunc: cancel,
+		ctx:        ctx,
+	}
+
+	go s.deleteExpired()
+	return s, nil
+}
+
+func (s *expirableStore) shutdown() error {
+	s.cancelFunc()
+	return nil
+}
+
+func (s *expirableStore) deleteExpired() {
+	ticker := time.NewTicker(s.bucketTTL)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-s.ctx.Done():
+			return
+		case <-ticker.C:
+			s.emptyExpiredBucket()
+		}
+	}
+}
+
+// TODO: document this
+func (s *expirableStore) fetch(id string, limit *Limit) (*Quota, error) {
+	select {
+	case <-s.ctx.Done():
+		return nil, ErrStopped
+	default:
+		// continue
+	}
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	key := getKey(limit.Resource, limit.Action, string(limit.Per), id)
+
+	e, ok := s.items[key]
+	switch {
+	case !ok:
+		e = s.pool.Get().(*entry)
+		e.key = key
+		e.value.reset(limit)
+		if err := s.add(e); err != nil {
+			s.pool.Put(e)
+			return nil, err
+		}
+		s.addToBucket(e)
+	case e.value.Expired():
+		s.removeFromBucket(e)
+		e.value.reset(limit)
+		s.addToBucket(e)
+	}
+
+	return e.value, nil
+}
+
+func (s *expirableStore) add(e *entry) error {
+	if _, ok := s.items[e.key]; !ok && len(s.items) >= s.maxSize {
+		return ErrLimiterFull
+	}
+	s.items[e.key] = e
+	return nil
+}
+
+func (s *expirableStore) addToBucket(e *entry) {
+	e.bucket = (int(e.value.limit.Period/s.bucketTTL) + s.nextBucketToExpire) % s.numberBuckets
+	s.buckets[e.bucket].entries[e.key] = e
```

Comment:
looks a little racey.

Reply:
Same here, this is only called by fetch which grabs the lock.